Fix the MathType editor's position on the screen

TylerZeroMaster opened the pull request:
part of openstax/ce#2205

If multiple instances are opened, they are stacked vertically.

This also moves the math source editor (wrs_code) in the same manner.

Multiple h5p editor instances can be opened at once. Each h5p editor instance has its own iframe.

The custom viewportUpdate event is sent to each iframe.

Consequently, when multiple h5p editors are opened and each has one or more MathType editor instances open, all MathType editors should be updated within their respective iframes.

Additionally, the MathType editor instances should remain contained within their respective iframes (see `top` in addons.js).

## Demo

https://github.com/openstax/vscode-h5p/assets/33585550/1f4eda15-ceb4-47ca-8c24-d1ceeadb1633

TylerZeroMaster committed:
Fix the MathType editor's position on the screen

If multiple instances are opened, they are stacked vertically.

This also moves the math source editor (wrs_code) in the same manner.

Multiple h5p editor instances can be opened at once. Each h5p editor instance
has its own iframe.

The custom viewportUpdate event is sent to each iframe.

Consequently, when multiple h5p editors are opened and each has one or more
MathType editor instances open, all MathType editors should be updated within
their respective iframes.

Additionally, the MathType editor instances should remain contained within
their respective iframes (see `top` in addons.js).

philschatz commented on server/static/editor-plugins/addons.js:
```diff
@@ -1,10 +1,99 @@
+        const boundingRects = visibleInstances
+          .map((dialog) => dialog.getBoundingClientRect())
+        const instanceHeights = boundingRects.map((rect) => rect.height)
+        const totalHeight = instanceHeights
```
Why sum up all the heights of all the mathType instances? I thought there was at most one dialog open per page

TylerZeroMaster replied:
I thought so too until I realized two things:
1. There **_can_** be more than one instance of MathType open at a time. Prior to this change that was unclear because the dialogs would stack up in one place.
2. There is a source editor that should be moved too.

The video demos both of these cases if you want to see.

philschatz commented on server/static/editor-plugins/addons.js:
```diff
@@ -1,10 +1,99 @@
+  }
+
+  const isMathTypeNode = (node) =>
+    node.id === 'wrs_code' || node.id?.startsWith('wrs_modal_dialogContainer')
```
The selector below uses `.wrs_modal_dialogContainer` but this checks an ID... why use different techniques for matching the element?

TylerZeroMaster replied:
The id of each dialog is like `wrs_modal_dialogContainer[0]`, `wrs_modal_dialogContainer[1]`, etc.

I agree that it is a little confusing that the selector uses class while the function uses ID and will change this to `node.classList?.contains('wrs_modal_dialogContainer')` instead.

philschatz commented on server/static/editor-plugins/addons.js:
```diff
@@ -1,10 +1,99 @@
+    );
+    if (waitingForUpdate || visibleInstances.length === 0) return;
+    waitingForUpdate = true;
+    requestAnimationFrame(() => {
```
Maybe `waitingForUpdate` and `shouldUpdate` could be combined into one variable named something like `isUpdatePending`? It seems like the intent is to update once after mutations have occurred.

If so, then it seems `requestAnimationFrame` only needs to run if it's not already queued up

TylerZeroMaster replied:
~Part of me wants to keep the observer as removed from `setMathTypePosition` as possible because I think `setMathTypePosition` could be repurposed into a more generic `update` function that could deal with positioning a variety of dialogs. Hopefully we will not need to position more elements, but I like to keep options open.~

~The `shouldUpdate` in the observer is being used to determine if there is a reason to call `setMathTypePosition`. It's a constant because I wanted to avoid nesting it inside the `if`.~

This comment convinced me that I should finish the "evergreen" approach instead of making something singularly aimed at the mathtype editor window. Granted, I am not going to add more complex tiling logic yet because I am not sure it will be required.

philschatz reviewed (approved):
It looks good! And I think that sending the position from the page to the iFrame is the ~~least-awkward~~ _best_ way to go!

I just had a couple questions/thoughts

TylerZeroMaster committed:
Use classList instead of id in isMathTypeNode

TylerZeroMaster committed:
Separate handling of mathtype dialogs and viewport updates

- ViewportUpdateHandler can position arbitrary elements
- MathTypeInstanceHandler watches for changes in mathtype instances
and triggers updates when necessary
- Can add more elements if needed by following a similar pattern
